app/controllers/predicates/AuthorisedAction.scala

```diff
@@ -99,35 +108,60 @@ class AuthorisedAction @Inject()()(implicit val authConnector: AuthConnector,
-      .withDelegatedAuthRule(agentDelegatedAuthRuleKey)
+      .withIdentifier(EnrolmentIdentifiers.individualId, mtdId)
+      .withDelegatedAuthRule(DelegatedAuthRules.agentDelegatedAuthRule)
+
+  private[predicates] def secondaryAgentPredicate(mtdId: String): Predicate =
+    Enrolment(EnrolmentKeys.SupportingAgent)
+      .withIdentifier(EnrolmentIdentifiers.individualId, mtdId)
+      .withDelegatedAuthRule(DelegatedAuthRules.supportingAgentDelegatedAuthRule)
 
+  private[predicates] def agentAuthentication[A](block: User[A] => Future[Result], mtdItId: String)
+                                                (implicit request: Request[A], hc: HeaderCarrier): Future[Result] = {
     authorised(agentAuthPredicate(mtdItId))
       .retrieve(allEnrolments) { enrolments =>
+        populateAgent(block, mtdItId, None, enrolments)
+      }.recoverWith(agentRecovery(block, mtdItId))
+  }
 
-        enrolmentGetIdentifierValue(EnrolmentKeys.Agent, EnrolmentIdentifiers.agentReference, enrolments) match {
-          case Some(arn) =>
-            block(User(mtdItId, Some(arn), "", optionalSessionId))
-          case None =>
-            logger.info("[AuthorisedAction][agentAuthentication] Agent with no HMRC-AS-AGENT enrolment.")
-            unauthorized
+  private def agentRecovery[A](block: User[A] => Future[Result], mtdItId: String)
+                              (implicit request: Request[A], hc: HeaderCarrier): PartialFunction[Throwable, Future[Result]] = {
+    case _: NoActiveSession =>
+      logger.info(s"[AuthorisedAction][agentAuthentication] - No active session.")
+      unauthorized
+    case _: AuthorisationException =>
+      if (appConfig.emaSupportingAgentsEnabled) {
+        authorised(secondaryAgentPredicate(mtdItId))
+          .retrieve(allEnrolments) { enrolments =>
+            populateAgent(block, mtdItId, None, enrolments)
+          }.recoverWith {
+            case _: AuthorisationException =>
+              logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have delegated authority for Client.")
+              unauthorized
+          }
+      } else {
+        logger.info(s"[AuthorisedAction][agentAuthentication] - Agent does not have secondary delegated authority for Client.")
+        unauthorized
+      }
+  }
+
+  private def populateAgent[A](block: User[A] => Future[Result],
+                               mtdItId: String,
+                               nino: Option[String],
+                               enrolments: Enrolments)(implicit request: Request[A], hc: HeaderCarrier): Future[Result] = {
+    enrolmentGetIdentifierValue(EnrolmentKeys.Agent, EnrolmentIdentifiers.agentReference, enrolments) match {
+      case Some(arn) =>
+        sessionIdFrom(request, hc).fold {
+          logger.info(s"[AuthorisedAction][agentAuthentication] - No session id in request.")
+          unauthorized
+        } { sessionId =>
+          block(User(mtdItId, Some(arn), nino.getOrElse(""), sessionId))
```

Comment:
I think we wanted to add `isSecondaryAgent` as a flag to the `User` model. So it's probably worth doing that now as part of this change.